xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,274 @@
+      symbol: input.symbol,
+      last: send_last
+    });
+    if (send_last) {
+      zero!<RunLengthDecoderState>()
+    } else {
+      RunLengthDecoderState {
+        symbol: input.symbol,
+        count: next_count,
+        last: input.last,
+      }
+    }
+  }
+}
+
+
+// RLE decoder specialization for the codegen
+proc RunLengthDecoder32 {
+  init {()}
+
+  config (
+    input_r: chan<DecInData<32, 2>> in,
+    output_s: chan<DecOutData<32>> out,
+  ) {
+    spawn RunLengthDecoder<u32:32, u32:2>(input_r, output_s);
+    ()
+  }
+
+  next (tok: token, state: ()) {
+    ()
+  }
+}
+
+// Tests
+
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH  = u32:32;
+
+type TestSymbol     = bits[TEST_SYMBOL_WIDTH];
+type TestCount      = bits[TEST_COUNT_WIDTH];
+type TestStimulus   = (TestSymbol, TestCount);
+type TestDecInData  = DecInData<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>;
+type TestDecOutData = DecOutData<TEST_SYMBOL_WIDTH>;
+
+// Check that RLE decoder will create 2 `last` output packets,
+// when 2 `last` input packets were consumed.
+#[test_proc]
+proc RunLengthDecoderLastAfterLastTest {
+  terminator: chan<bool> out;            // test termination request
+  dec_input_s: chan<TestDecInData> out;
+  dec_output_r: chan<TestDecOutData> in;
+
+  init {()}
+
+  config(terminator: chan<bool> out) {
+    let (dec_input_s, dec_input_r)   = chan<TestDecInData>;
+    let (dec_output_s, dec_output_r) = chan<TestDecOutData>;
+
+    spawn RunLengthDecoder<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>(
+        dec_input_r, dec_output_s);
+    (terminator, dec_input_s, dec_output_r)
+  }
+
+  next(tok: token, state: ()) {
+    let LastAfterLastTestStimuli: TestDecInData[2] =[
+      TestDecInData {
+        symbol: TestSymbol:0x1,
+        count: TestCount:0x1,
+        last:true
+      },
+      TestDecInData {
+        symbol: TestSymbol:0x2,
+        count: TestCount:0x1,
+        last:true
+      },
+    ];
+    let tok = for ((counter, stimulus), tok):
+        ((u32, TestDecInData) , token)
+        in enumerate(LastAfterLastTestStimuli) {
+      let tok = send(tok, dec_input_s, stimulus);
+      trace_fmt!("Sent {} stimuli, symbol: 0x{:x}, count:{}, last: {}",
+          counter + u32:1, stimulus.symbol, stimulus.count, stimulus.last);
+      (tok)
+    }(tok);
+    let LastAfterLastTestOutputs: TestDecOutData[2] = [
+      TestDecOutData{symbol: TestSymbol: 0x1, last: true},
+      TestDecOutData{symbol: TestSymbol: 0x2, last: true},
+    ];
+    let tok = for ((counter, output), tok):
+        ((u32, TestDecOutData) , token)
+        in enumerate(LastAfterLastTestOutputs) {
+      let (tok, dec_output) = recv(tok, dec_output_r);
+      trace_fmt!(
+          "Received {} transactions, symbol: 0x{:x}, last: {}",
+          counter + u32:1, dec_output.symbol, dec_output.last
+      );
+      assert_eq(dec_output, output);
+      (tok)
+    }(tok);
+    send(tok, terminator, true);
+    ()
+  }
+}
+
+// Check RLE decoder on longer transaction
```

Comment:
> "longer transaction"

It's not really "longer" since there is no more simple tests :)

Reply:
Removed "longer" from the comment.